File: apps/android/app/src/main/kotlin/dev/hyphen/android/transport/SessionLiveness.kt
package dev.hyphen.android.transport

/**
 * Receive-silence watchdog (protocol v0 §4, HYP-M2-012): DEGRADED after
 * `missThreshold` heartbeat intervals with no envelope received (the
 * spec's "two consecutive missed heartbeats" — peers send heartbeats
 * every interval, so silence is the observable). Any received envelope
 * recovers to HEALTHY. Pure logic; time is injected.
 */
class HeartbeatMonitor(
    private val intervalMs: Long,
    private val missThreshold: Int = 2,
    startedAtMs: Long,
    private val onStateChange: (State) -> Unit,
) {
    enum class State { HEALTHY, DEGRADED }

    var state: State = State.HEALTHY
        private set

    private var lastReceivedAtMs = startedAtMs

    @Synchronized
    fun envelopeReceived(nowMs: Long) {
        lastReceivedAtMs = nowMs
        if (state == State.DEGRADED) {
            state = State.HEALTHY
            onStateChange(state)
        }
    }

    @Synchronized
    fun tick(nowMs: Long) {
        if (state == State.HEALTHY && nowMs - lastReceivedAtMs > intervalMs * missThreshold) {
            state = State.DEGRADED
            onStateChange(state)
        }
    }
}

/**
 * Tracks sent `requiresAck` envelopes (protocol v0 §3): if no ack arrives
 * within `timeoutMs`, the sender treats it as `protocol/ack-timeout`.
 * Each timeout fires exactly once. Pure logic; time is injected.
 */
class AckTracker(
    private val timeoutMs: Long,
    private val onTimeout: (messageId: String) -> Unit,
) {
    private val pending = LinkedHashMap<String, Long>()

    @Synchronized
    fun registerSent(messageId: String, nowMs: Long) {
        pending[messageId] = nowMs
    }

    @Synchronized
    fun ackReceived(ackOf: String): Boolean = pending.remove(ackOf) != null

    fun tick(nowMs: Long) {
        val expired: List<String>
        synchronized(this) {
            expired = pending.filterValues { sentAt -> nowMs - sentAt > timeoutMs }.keys.toList()
            expired.forEach(pending::remove)
        }
        expired.forEach(onTimeout)
    }

    @Synchronized
    fun pendingCount(): Int = pending.size
}

File: apps/macos/Sources/HyphenTransport/SessionLiveness.swift
import Foundation

/// Receive-silence watchdog (protocol v0 §4, HYP-M2-012): degraded after
/// `missThreshold` heartbeat intervals with no envelope received. Any
/// received envelope recovers to healthy. Pure logic; time is injected.
/// Not thread-safe by itself — the session confines it to one queue.
public final class HeartbeatMonitor {

    public enum State: Equatable {
        case healthy
        case degraded
    }

    public private(set) var state: State = .healthy

    private let intervalMs: Int64
    private let missThreshold: Int64
    private var lastReceivedAtMs: Int64
    private let onStateChange: (State) -> Void

    public init(
        intervalMs: Int64,
        missThreshold: Int64 = 2,
        startedAtMs: Int64,
        onStateChange: @escaping (State) -> Void
    ) {
        self.intervalMs = intervalMs
        self.missThreshold = missThreshold
        self.lastReceivedAtMs = startedAtMs
        self.onStateChange = onStateChange
    }

    public func envelopeReceived(nowMs: Int64) {
        lastReceivedAtMs = nowMs
        if state == .degraded {
            state = .healthy
            onStateChange(state)
        }
    }

    public func tick(nowMs: Int64) {
        if state == .healthy, nowMs - lastReceivedAtMs > intervalMs * missThreshold {
            state = .degraded
            onStateChange(state)
        }
    }
}

/// Tracks sent `requiresAck` envelopes (protocol v0 §3): no ack within
/// `timeoutMs` is `protocol/ack-timeout`, fired exactly once per id.
public final class AckTracker {

    private var pending: [String: Int64] = [:]
    private let timeoutMs: Int64
    private let onTimeout: (String) -> Void

    public init(timeoutMs: Int64, onTimeout: @escaping (String) -> Void) {
        self.timeoutMs = timeoutMs
        self.onTimeout = onTimeout
    }

    public func registerSent(messageId: String, nowMs: Int64) {
        pending[messageId] = nowMs
    }

    @discardableResult
    public func ackReceived(_ ackOf: String) -> Bool {
        pending.removeValue(forKey: ackOf) != nil
    }

    public func tick(nowMs: Int64) {
        let expired = pending.filter { nowMs - $0.value > timeoutMs }.keys.sorted()
        for messageId in expired {
            pending.removeValue(forKey: messageId)
            onTimeout(messageId)
        }
    }

    public var pendingCount: Int { pending.count }
}
